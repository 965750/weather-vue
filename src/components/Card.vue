<template>
    <div class="card mt-5 pa-3" v-if="getCard">
        <p class="card__city mb-0 text-center">
            {{ getCard.city.name }}
            <span class="card__country overline">
                {{ getCard.city.country }}
            </span>
        </p>
        <p class="my-0 text-center caption">
            {{ currentDate }}
        </p>
        <SunVisual :clouds="extendedWeather.averages.cloudsTotal" />

        <DayPeriods :periods="extendedWeather.periods"/>
    </div>
</template>

<script>
import { mapGetters } from 'vuex';
import SunVisual from './SunVisual.vue';
import DayPeriods from './DayPeriods.vue';

export default {
  name: 'Card',
  components: {
    SunVisual,
    DayPeriods,
  },
  computed: {
    ...mapGetters(['getCard']),
    currentDate() {
      const time = new Date();

      const day = this.$options.filters.shortDaysOfWeek(time.getDay() - 1);

      return `${day}, ${time.getHours()}:${time.getMinutes()}`;
    },
    extendedWeather() {
      let timeCounter = 0;
      const weather = {
        tempRange: {
          minTemp: 0,
          maxTemp: 0,
        },
        averages: {
          cloudsTotal: 0,
          humidityTotal: 0,
          tempTotal: 0,
          windTotal: 0,
        },
        periods: {
          totals: {
            morning: 0,
            day: 0,
            night: 0,
          },
          counters: {
            morning: 0,
            day: 0,
            night: 0,
          },
        },
      };

      this.getCard.weatherList.forEach((current) => {
        const time = new Date(current.time * 1000).getHours() + 1;

        console.log(current);
        if (new Date(current.time).getDay() === this.currentDay) {
          timeCounter += 1;
          // averages
          weather.averages.cloudsTotal += current.clouds;
          weather.averages.humidityTotal += current.main.humidity;
          weather.averages.tempTotal += current.main.temp;
          weather.averages.windTotal += current.wind;
          // temp range
          if (current.main.temp_min < weather.tempRange.minTemp || !weather.tempRange.minTemp) {
            weather.tempRange.minTemp = current.main.temp_min;
          }

          if (current.main.temp_max > weather.tempRange.maxTemp || !weather.tempRange.maxTemp) {
            weather.tempRange.maxTemp = current.main.temp_max;
          }
          // day periods
          if (time >= 3 && time <= 12) {
            weather.periods.totals.morning += current.main.temp;
            weather.periods.counters.morning += 1;
          }

          if (time >= 9 && time <= 18) {
            weather.periods.totals.day += current.main.temp;
            weather.periods.counters.day += 1;
          }

          if (time >= 21 || time <= 6) {
            weather.periods.totals.night += current.main.temp;
            weather.periods.counters.night += 1;
          }
        }
      });

      return {
        ...weather,
        averages: {
          cloudsTotal: weather.averages.cloudsTotal / timeCounter,
          humidityTotal: weather.averages.humidityTotal / timeCounter,
          tempTotal: weather.averages.tempTotal / timeCounter,
          windTotal: weather.averages.windTotal / timeCounter,
        },
        periods: {
          morning: weather.periods.totals.morning / weather.periods.counters.morning,
          day: weather.periods.totals.day / weather.periods.counters.day,
          night: weather.periods.totals.night / weather.periods.counters.night,
        },
      };
    },
    currentDay() {
      return new Date(this.getCard.weatherList[0].time).getDay();
    },
  },
};
</script>

<style lang="scss">
@import '../assets/main';

.card {
    width: 300px;
    height: 500px;
    border: 1px solid red;
    margin: auto;
    border-radius: 10px;
    background: cornflowerblue;
    color: #FFF;

    &__country {
        color: $aluminium;
    }
}
</style>
