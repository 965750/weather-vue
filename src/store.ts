import Vue from 'vue';
import Vuex from 'vuex';
import axios from 'axios';
import config from '@/config';

import mutations from './mutations';
import getters from './getters';

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    lang: 'pl',
    card: null,
    error: null,
  },
  actions: {
    citySearch({ commit }, city: string) {
      axios
        .get(`https://api.openweathermap.org/data/2.5/forecast?q=${city}&cnt=${config.cnt}&APPID=${config.apiKey}`)
        .then((res) => {
          commit('setCard', res.data);
        })
        .catch((err) => {
          commit('setCard__Error', err);
        });
    },
  },
  mutations,
  getters,
});
