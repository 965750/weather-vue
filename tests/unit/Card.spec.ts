import {
  shallowMount,
  createLocalVue,
} from '@vue/test-utils';
import VueI18n from 'vue-i18n';
import Vuex from 'vuex';
import Card from '@/components/Card.vue';
import filters from '@/myFilters';
import store from '@/store';
import messages from '@/locale/languages';

const localVue = createLocalVue();
localVue.use(VueI18n);

const initialStore = {
  getters: {
    getCard: () => ({
      city: {
        name: 'Berlin',
        country: 'de',
      },
      weatherList: [
        {
          clouds: 10,
          time: 16123124,
          wind: 4,
          rain: 2,
          main: {
            humidity: 70,
            temp: 288,
            temp_min: 282,
            temp_max: 298,
          },
        },
        {
          clouds: 11,
          time: 16123124,
          wind: 5,
          rain: 2,
          main: {
            humidity: 76,
            temp: 286,
            temp_min: 281,
            temp_max: 294,
          },
        },
      ],
    }),
  },
};

const i18n = new VueI18n({
  locale: store.getters.getLanguage,
  messages,
});

filters.forEach((filter) => {
  localVue.filter(filter.name, filter.handler);
});

describe('Card.vue', () => {
  const createWrapper = (mockStore = initialStore) => shallowMount(Card, {
    store: new Vuex.Store(mockStore),
    i18n,
    localVue,
  });

  it('should match initial snapshot', () => {
    const wrapper = createWrapper();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should update temp_min in extendedWeather', () => {
    const wrapper = createWrapper({
      getters: {
        getCard: () => ({
          city: {
            name: 'Berlin',
            country: 'de',
          },
          weatherList: [
            {
              clouds: 10,
              time: 16123124,
              wind: 4,
              rain: 2,
              main: {
                humidity: 70,
                temp: 288,
                temp_min: 282,
                temp_max: 298,
              },
            },
            {
              clouds: 11,
              time: 16123124,
              wind: 5,
              rain: 2,
              main: {
                humidity: 76,
                temp: 286,
                temp_min: 285,
                temp_max: 299,
              },
            },
          ],
        }),
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should update pass NO raid in extendedWeather', () => {
    const wrapper = createWrapper({
      getters: {
        getCard: () => ({
          city: {
            name: 'Berlin',
            country: 'de',
          },
          weatherList: [
            {
              clouds: 10,
              time: 16123124,
              wind: 4,
              rain: 0,
              main: {
                humidity: 70,
                temp: 288,
                temp_min: 282,
                temp_max: 298,
              },
            },
          ],
        }),
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should update set valid dominant value as mostCommon temp', () => {
    const wrapper = createWrapper({
      getters: {
        getCard: () => ({
          city: {
            name: 'Berlin',
            country: 'de',
          },
          weatherList: [
            {
              clouds: 10,
              time: 16123124,
              wind: 4,
              rain: 2,
              main: {
                humidity: 70,
                temp: 288,
                temp_min: 282,
                temp_max: 298,
              },
            },
            {
              clouds: 11,
              time: 16123124,
              wind: 5,
              rain: 2,
              main: {
                humidity: 76,
                temp: 288,
                temp_min: 285,
                temp_max: 299,
              },
            },
            {
              clouds: 11,
              time: 16123124,
              wind: 5,
              rain: 2,
              main: {
                humidity: 76,
                temp: 283,
                temp_min: 285,
                temp_max: 299,
              },
            },
          ],
        }),
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
