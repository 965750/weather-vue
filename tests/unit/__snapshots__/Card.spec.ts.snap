// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue should match initial snapshot 1`] = `
<div
  class="card mt-5 px-3 py-2"
>
  <p
    class="card__city mb-0 text-center"
  >
    
        Berlin
        
    <span
      class="card__country overline"
    >
      
            de
        
    </span>
  </p>
   
  <p
    class="card__date my-0 text-center caption"
  >
    
        
    
  </p>
   
  <sunvisual-stub
    clouds="10.5"
    rain="2"
  />
   
  <p
    class="mainTemp text-center"
  >
    <span
      class="display-2 font-weight-bold"
    >
      
            15
        
    </span>
     
    <span
      class="mainTemp__celsius"
    >
      ℃
    </span>
  </p>
   
  <statistics-stub
    class="card__statistics"
    weather="[object Object]"
  />
   
  <timeperiods-stub
    class="card__timePeriods"
    periods="[object Object],[object Object]"
  />
</div>
`;

exports[`Card.vue should update pass NO raid in extendedWeather 1`] = `
<div
  class="card mt-5 px-3 py-2"
>
  <p
    class="card__city mb-0 text-center"
  >
    
        Berlin
        
    <span
      class="card__country overline"
    >
      
            de
        
    </span>
  </p>
   
  <p
    class="card__date my-0 text-center caption"
  >
    
        
    
  </p>
   
  <sunvisual-stub
    clouds="10"
  />
   
  <p
    class="mainTemp text-center"
  >
    <span
      class="display-2 font-weight-bold"
    >
      
            15
        
    </span>
     
    <span
      class="mainTemp__celsius"
    >
      ℃
    </span>
  </p>
   
  <statistics-stub
    class="card__statistics"
    weather="[object Object]"
  />
   
  <timeperiods-stub
    class="card__timePeriods"
    periods="[object Object]"
  />
</div>
`;

exports[`Card.vue should update set valid dominant value as mostCommon temp 1`] = `
<div
  class="card mt-5 px-3 py-2"
>
  <p
    class="card__city mb-0 text-center"
  >
    
        Berlin
        
    <span
      class="card__country overline"
    >
      
            de
        
    </span>
  </p>
   
  <p
    class="card__date my-0 text-center caption"
  >
    
        
    
  </p>
   
  <sunvisual-stub
    clouds="10.666666666666666"
    rain="2"
  />
   
  <p
    class="mainTemp text-center"
  >
    <span
      class="display-2 font-weight-bold"
    >
      
            15
        
    </span>
     
    <span
      class="mainTemp__celsius"
    >
      ℃
    </span>
  </p>
   
  <statistics-stub
    class="card__statistics"
    weather="[object Object]"
  />
   
  <timeperiods-stub
    class="card__timePeriods"
    periods="[object Object],[object Object],[object Object]"
  />
</div>
`;

exports[`Card.vue should update temp_min in extendedWeather 1`] = `
<div
  class="card mt-5 px-3 py-2"
>
  <p
    class="card__city mb-0 text-center"
  >
    
        Berlin
        
    <span
      class="card__country overline"
    >
      
            de
        
    </span>
  </p>
   
  <p
    class="card__date my-0 text-center caption"
  >
    
        
    
  </p>
   
  <sunvisual-stub
    clouds="10.5"
    rain="2"
  />
   
  <p
    class="mainTemp text-center"
  >
    <span
      class="display-2 font-weight-bold"
    >
      
            15
        
    </span>
     
    <span
      class="mainTemp__celsius"
    >
      ℃
    </span>
  </p>
   
  <statistics-stub
    class="card__statistics"
    weather="[object Object]"
  />
   
  <timeperiods-stub
    class="card__timePeriods"
    periods="[object Object],[object Object]"
  />
</div>
`;
